### ISSUE TITLE
Python kernel dies without error message when assigning quaternion to numerical numpy array

### ISSUE THREAD

**Issue by Plankton555** [moble/quaternion#116]
**Describe the bug**
Python kernel/process dies without error message when trying to assign a quaternion as an element in a numerical numpy array.

**To Reproduce**
```python
import numpy as np
import quaternion

arr = np.zeros(5)
arr[0] = np.quaternion(1, 0, 0, 0)
```

**Expected behavior**
I did NOT expect the above code to successfully run, but I did expect an error message of some sort. Something that could point me in the the direction of the problem so I can fix the faulty code I introduced by mistake.

**Environment (please complete the following information):**
 - OS: Windows 10 Pro, version 1803 (OS build 17134.829)
 - Installation method: pip and virtualenv
 - Numpy version: '1.16.3'
 - Quaternion version: '2019.3.21.14.22.55'

**Additional context**
N/A

**Comment by Plankton555** [comment]
I realize now afterwards that perhaps this can't be fixed in the quaternion package, but must be handled in numpy?

**Comment by moble** [comment]
Unless there's something I'm missing, I think this must be a problem in numpy.  Comparing with numpy's own [`rational` test type](https://github.com/numpy/numpy/blob/464f79eb1d05bf938d16b49da1c39a4e02506fa3/numpy/core/src/umath/_rational_tests.c.src), I don't see many substantial differences.  However, a similar test assigning rational numbers to the array elements works.  The only important differences I see are that `rational`s can be cast to other dtypes (including floats).  It doesn't generally make any sense to cast a quaternion to other dtypes (including floats), but I tested it anyway [by adding (silly, but functional) casts](https://github.com/moble/quaternion/commit/ef3e9673acb9fecdb6faf22e390292d6bd220260) to this code, just to see if casts were being called, and they weren't.

@eric-wieser Is there something I'm supposed to be doing in this package to avoid this segfault?

**Comment by moble** [comment]
After searching around some more for any similar issues in numpy, I found [this one](https://github.com/numpy/numpy/issues/10693), and realized that the problem apparently boils down to just trying to convert a quaternion to a float (or int).  That is, you get a segfault with this code too:
```python
import numpy as np
import quaternion

s = float(np.quaternion(1, 0, 0, 0))
```
That issue (along with [this related one](https://github.com/numpy/numpy/issues/9972)) made it clear how to fix the problem.  So I've added some code that will detect when the user is trying to convert a quaternion to a float (or int; or long, oct, or hex in python 2).  The original code sample above should now produce an error that looks something like this:
```python
Traceback (most recent call last):
  File "test.py", line 5, in <module>
    arr[0] = np.quaternion(1, 0, 0, 0)
TypeError: Cannot convert quaternion to float
```
I've also added tests to ensure that this happens in future versions.
